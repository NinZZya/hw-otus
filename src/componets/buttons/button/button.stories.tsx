import React from "react";
// also exported from "@storybook/react" if you can deal with breaking changes in 6.1
import { Meta } from "@storybook/react/types-6-0";

import { Button } from "./button";

export default {
  title: "Example/buttons/Button",
  component: Button,
} as Meta;

export const Simple: React.FC = () => <Button>Button</Button>;
