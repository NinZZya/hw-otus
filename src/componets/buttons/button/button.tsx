import React from "react";
/** @jsx jsx */
import { jsx, css, useTheme } from "@emotion/react";
import { SerializedStyles } from "@emotion/utils/types";

interface Theme {
  color?: {
    text: string;
    primary: string;
    border: string;
    shadowBorder: string;
  };
}

type ButtonProp = React.DetailedHTMLProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
>;

const DefaultColor = {
  TEXT: "#ffffff",
  BACKGROUND: "#3700b3",
  SHADOW_BORDER: "rgba(255, 255, 255, 0.1)",
};

const getButtonStyle = (theme: Theme): SerializedStyles => css`
  display: inline-block;
  padding: 1rem;
  font-weight: 700;
  border-radius: 0.3em;
  cursor: pointer;
  line-height: 1;
  text-decoration: none;
  border: 1px solid transparent;
  color: ${theme.color ? theme.color.text : DefaultColor.TEXT};
  background-color: ${theme.color
    ? theme.color.primary
    : DefaultColor.BACKGROUND};

  &:focus,
  &:hover {
    outline: none;
    transperent: 0.4;
    border: 1px solid
      ${theme.color ? theme.color.shadowBorder : DefaultColor.SHADOW_BORDER};
    box-shadow: 0px 0px 5px 1px
      ${theme.color ? theme.color.shadowBorder : DefaultColor.SHADOW_BORDER};
  }

  &:disabled {
    background-color: ${theme.color
      ? theme.color.shadowBorder
      : DefaultColor.SHADOW_BORDER};
  }
`;

export const Button: React.FC<ButtonProp> = (props) => {
  const theme = useTheme();

  return <button css={getButtonStyle(theme)} {...props} />;
};
