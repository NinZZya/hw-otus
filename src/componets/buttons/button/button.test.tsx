import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import { Button } from "./button";

configure({ adapter: new Adapter() });

describe(`Button render coorectly`, () => {
  it(`Button without props render coorectly`, () => {
    const tree = renderer.create(<Button />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Button with text render coorectly`, () => {
    const tree = renderer.create(<Button>Button</Button>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Button was clicked`, () => {
    const onClick = jest.fn();
    const button = shallow(<Button onClick={onClick} />);

    button.simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
