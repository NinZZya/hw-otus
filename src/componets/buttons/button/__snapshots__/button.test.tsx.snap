// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render coorectly Button with text render coorectly 1`] = `
<button
  css={
    Object {
      "map": undefined,
      "name": "l4n4cz",
      "next": undefined,
      "styles": "
  display: inline-block;
  padding: 1rem;
  font-weight: 700;
  border-radius: 0.3em;
  cursor: pointer;
  line-height: 1;
  text-decoration: none;
  border: 1px solid transparent;
  color: #ffffff;
  background-color: #3700b3;

  &:focus,
  &:hover {
    outline: none;
    transperent: 0.4;
    border: 1px solid
      rgba(255, 255, 255, 0.1);
    box-shadow: 0px 0px 5px 1px
      rgba(255, 255, 255, 0.1);
  }

  &:disabled {
    background-color: rgba(255, 255, 255, 0.1);
  }
",
      "toString": [Function],
    }
  }
>
  Button
</button>
`;

exports[`Button render coorectly Button without props render coorectly 1`] = `
<button
  css={
    Object {
      "map": undefined,
      "name": "l4n4cz",
      "next": undefined,
      "styles": "
  display: inline-block;
  padding: 1rem;
  font-weight: 700;
  border-radius: 0.3em;
  cursor: pointer;
  line-height: 1;
  text-decoration: none;
  border: 1px solid transparent;
  color: #ffffff;
  background-color: #3700b3;

  &:focus,
  &:hover {
    outline: none;
    transperent: 0.4;
    border: 1px solid
      rgba(255, 255, 255, 0.1);
    box-shadow: 0px 0px 5px 1px
      rgba(255, 255, 255, 0.1);
  }

  &:disabled {
    background-color: rgba(255, 255, 255, 0.1);
  }
",
      "toString": [Function],
    }
  }
/>
`;
